src/main/groovy/lang/NumberRange.java

```diff
@@ -0,0 +1,603 @@
+        }
+
+        // Performance detail:
+        // not using get(fromIndex), get(toIndex) in the following to avoid stepping over elements twice
+        final Iterator<Comparable> iter = new StepIterator(this, stepSize);
+
+        Comparable value = iter.next();
+        int i = 0;
+        for (; i < fromIndex; i++) {
+            if (!iter.hasNext()) {
+                throw new IndexOutOfBoundsException("Index: " + i + " is too big for range: " + this);
+            }
+            value = iter.next();
+        }
+        final Comparable fromValue = value;
+        for (; i < toIndex - 1; i++) {
+            if (!iter.hasNext()) {
+                throw new IndexOutOfBoundsException("Index: " + i + " is too big for range: " + this);
+            }
+            value = iter.next();
+        }
+        final Comparable toValue = value;
+
+        return new NumberRange(comparableNumber(fromValue), comparableNumber(toValue), comparableNumber(stepSize), true);
+    }
+
+    @Override
+    public String toString() {
+        return getToString(to.toString(), from.toString());
+    }
+
+    @Override
+    public String inspect() {
+        return getToString(InvokerHelper.inspect(to), InvokerHelper.inspect(from));
+    }
+
+    private String getToString(String toText, String fromText) {
+        String sep = inclusive ? ".." : "..<";
+        String base = reverse ? "" + toText + sep + fromText : "" + fromText + sep + toText;
+        return Integer.valueOf(1).equals(stepSize) ? base : base + ".by(" + stepSize + ")";
+    }
+
+    /**
+     * iterates over all values and returns true if one value matches.
+     * Also see containsWithinBounds.
+     */
+    @Override
+    public boolean contains(Object value) {
+        final Iterator it = new StepIterator(this, stepSize);
+        if (value == null) {
+            return false;
+        }
+        while (it.hasNext()) {
+            try {
+                if (compareEqual(value, it.next())) {
+                    return true;
+                }
+            } catch (ClassCastException e) {
+                return false;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void step(int numSteps, Closure closure) {
+        if (numSteps == 0 && compareTo(from, to) == 0) {
+            return; // from == to and step == 0, nothing to do, so return
+        }
+        final StepIterator iter = new StepIterator(this, multiply(numSteps, stepSize));
+        while (iter.hasNext()) {
+            closure.call(iter.next());
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<Comparable> iterator() {
+        // non thread-safe iterator
+        final Iterator<Comparable> innerIterator = new StepIterator(this, stepSize);
+        return new Iterator<Comparable>() {
+            @Override
+            public synchronized boolean hasNext() {
```

Comment:
To me it seems like a true thread-safe iterator would need to lock in `hasNext()` and unlock when it returns false (or throws) or after a subsequent call by the same lock-holding thread to `next()`.  Though I haven't test it, it appears with these synchronized methods that a thread could have `hasNext()` return `true` and a subsequent call to `next()` return `null` (though maybe should be `NoSuchElementException`).

But I still do not understand the benefit of having a thread safe version of the iterator that multiple threads can share and wonder if maybe those synchronized methods were put in by mistake (especially given the nearby comments about it not being thread safe).  Also, I don't see similar synchronization in the `IntRange` class.